### PR TITLE
Revamps PageCache to store data in one entry (#10506)

### PR DESCRIPTION
**Pull request by sergeymakinen**
| Q | A |
| --- | --- |
| Is bugfix? | yes |
| New feature? | yes |
| Breaks BC? | no |
| Tests pass? | yes |
| Fixed issues | #10506 |

This updated `PageCache` combines together `PageCache` and `FragmentCache` to provide caching in one entry.
### Features
- fully backwards compatible (incl. dynamic content)
- easy to be extended
- able to store/retrieve an additional data in `beforeCacheResponse` and `afterRestoreResponse` (_for example: I store Last-Modified in data and then retrieve it later in `HttpCache`_)
- all data is stored in one cache entry
- able to cancel caching by returning `false` in `beforeCacheResponse`

### CODE REVIEW

File: framework/filters/PageCache.php [hunk 1]
```diff
@@ -123,6 +124,12 @@ class PageCache extends ActionFilter
      * @since 2.0.4
      */
     public $cacheHeaders = true;
+    /**
+     * @var array a list of placeholders for embedding dynamic contents. This property
+     * is used internally to implement the content caching feature. Do not modify it.
+     * @since 2.0.10
+     */
+    public $dynamicPlaceholders;
```
Comment (samdark) [inline framework/filters/PageCache.php]: Should it be public?

**Comment by samdark** [comment]
@yiisoft/core-developers need additional review.

**Comment by samdark** [comment]
@cebe, @drscre, @kidol

File: framework/filters/PageCache.php (hunk 1 above)
Comment (sergeymakinen) [inline framework/filters/PageCache.php]: Yes, `View` uses it.

Comment (samdark) [inline framework/filters/PageCache.php]: Is there a way to avoid exposing it?

Comment (sergeymakinen) [inline framework/filters/PageCache.php]: Not sure… I guess it would require changes in 'View' and any other classes implementing a dynamic content.

**Review by samdark — APPROVED** [review]
It looks good to me but I need more reviews from other team members.

File: framework/filters/PageCache.php [hunk 2]
```diff
@@ -123,6 +124,12 @@ class PageCache extends ActionFilter
      * @since 2.0.4
      */
     public $cacheHeaders = true;
+    /**
+     * @var array a list of placeholders for embedding dynamic contents. This property
+     * is used internally to implement the content caching feature. Do not modify it.
+     * @since 2.0.11
```
Comment (SilverFire) [inline framework/filters/PageCache.php]: Add `@internal`

File: framework/filters/PageCache.php [hunk 3]
```diff
@@ -154,57 +161,74 @@ public function beforeAction($action)
...
+     * @return boolean|array whether to cache or not, return an array instead of `true` to store an additional data.
+     * @since 2.0.11
+     */
+    public function beforeCacheResponse()
+    {
+        return true;
+    }
+
+    /**
+     * This method is invoked right after restoring a response is finished (but before the response is sent).
```
Comment (SilverFire) [inline framework/filters/PageCache.php]: Isn't it better to change spelling like this:

`This method is invoked right after the response restoring is finished`

File: framework/filters/PageCache.php [hunk 4]
```diff
@@ -213,43 +237,71 @@ protected function restoreResponse($response, $data)
      */
     public function cacheResponse()
     {
-        $this->view->endCache();
-        $response = Yii::$app->getResponse();
+        array_pop($this->view->cacheStack);
+        $beforeResult = $this->beforeCacheResponse();
```
Comment (SilverFire) [inline framework/filters/PageCache.php]: Maybe `beforeResponse`?

File: framework/filters/PageCache.php [hunk 5]
```diff
@@ -213,43 +237,71 @@ protected function restoreResponse($response, $data)
...
+            $data[$name] = $response->{$name};
+        }
+        foreach (['headers', 'cookies'] as $name) {
+            $property = 'cache' . ucfirst($name);
+            if ($this->{$property} !== false) {
+                $all = $response->{$name}->toArray();
+                if (is_array($this->{$property})) {
+                    $filtered = [];
+                    foreach ($this->{$property} as $subName) {
+                        if ($name === 'headers') {
```
Comment (SilverFire) [inline framework/filters/PageCache.php]: Looks like a crutch

File: framework/filters/PageCache.php [hunk 6]
```diff
@@ -213,43 +237,71 @@ protected function restoreResponse($response, $data)
...
+                            $subName = strtolower($subName);
+                        }
+                        if (isset($all[$subName])) {
```
Comment (SilverFire) [inline framework/filters/PageCache.php]: Conditional nesting is too deep and it makes hard to understand what's going on here.
Could be split to a separate private method with self-documenting name.

File: framework/filters/PageCache.php [hunk 7]
```diff
@@ -213,43 +237,71 @@ protected function restoreResponse($response, $data)
...
+        if ($beforeResult === false) {
+            $content = ob_get_clean();
+            if (empty($this->view->cacheStack) && !empty($this->dynamicPlaceholders)) {
+                $content = $this->updateDynamicContent($content, $this->dynamicPlaceholders);
+            }
+            echo $content;
+            return;
+        }
+
+        $response = Yii::$app->response;
```
Comment (SilverFire) [inline framework/filters/PageCache.php]: The entire `cacheResponse` method does not fit on one screen.
Should be split for a smaller ones.

**Review by SilverFire — CHANGES_REQUESTED** [review]

**Comment by sergeymakinen** [comment]
@SilverFire Well, thanks for the review, I'm back and I slightly changed some parts according to your review.

And em… As 2.0.10 gets delayed more and more, may be we can get this into the 2.0.10 milestone, what do you think? :)

**Comment by SilverFire** [comment]
2.0.10 is delayed because we still have issues to resolve. Moving new issues back to 2.0.10 will delay it even more

File: framework/CHANGELOG.md [hunk 8]
```diff
@@ -68,6 +68,8 @@ Yii Framework 2 Change Log
 - Bug #12100: Fixed `yii\filters\HttpCache` was sending an empty Pragma header (sergeymakinen)
 - Bug #12107: Fixed REST Serializer to validate input for 'expand' and 'fields' parameter, which crashed on array input (njspok, cebe)
 - Bug #12143: Fixed `yii\db\BaseActiveRecord::updateAttributes()` change `isNewRecord` state for the new model (klimov-paul)
+- Enh: Added `beforeCacheResponse` and `afterRestoreResponse` to `yii\filters\PageCache` to be more easily extendable (sergeymakinen)
```
Comment (samdark) [inline framework/CHANGELOG.md]: Something is very wrong with changelog...

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/CHANGELOG.md (hunk 8 above)
Comment (sergeymakinen) [inline framework/CHANGELOG.md]: I will figure it out as soon as I get tests up & running. I hope it happens soon.

Comment (cebe) [inline framework/CHANGELOG.md]: best way to fix it would be to rebase against master.

**Comment by cebe** [comment]
I have stared to work on a test case for this class: #13124 that should be added before applying your changes to reduce the risk of unspotted breaks.

**Comment by sergeymakinen** [comment]
Yahoo! Travis is green now. The PR is ready to roll.

**Review by samdark — APPROVED** [review]

**Review by SilverFire — APPROVED** [review]

File: framework/filters/PageCache.php [hunk 9]
```diff
@@ -123,6 +124,13 @@ class PageCache extends ActionFilter
      * @since 2.0.4
      */
     public $cacheHeaders = true;
+    /**
+     * @var array a list of placeholders for embedding dynamic contents. This property
+     * is used internally to implement the content caching feature. Do not modify it.
+     * @internal
+     * @since 2.0.11
+     */
```
Comment (dynasource) [inline framework/filters/PageCache.php]: this naming is a bit confusing as it look like the placeholders are dynamic. I'd prefer something like "dynamicContentPlaceholders"

File: framework/filters/PageCache.php [hunk 10]
```diff
@@ -123,6 +124,13 @@ class PageCache extends ActionFilter
      * @since 2.0.4
      */
     public $cacheHeaders = true;
+    /**
+     * @var array a list of placeholders for embedding dynamic contents. This property
+     * is used internally to implement the content caching feature. Do not modify it.
+     * @internal
```
Comment (dynasource) [inline framework/filters/PageCache.php]: a public $variable which we don't want to modify? Shouldnt we use a setter than?

**Review by dynasource — CHANGES_REQUESTED** [review]

File: framework/filters/PageCache.php (hunk 10 above)
Comment (sergeymakinen) [inline framework/filters/PageCache.php]: It's a [convention](https://github.com/yiisoft/yii2/blob/master/framework/base/View.php#L363) I have to follow.

File: framework/filters/PageCache.php (hunk 9 above)
Comment (sergeymakinen) [inline framework/filters/PageCache.php]: Me too. But it's how [it works](#r92295170)…

I can improve it later with a getter/setter, some `DynamicContentInterface` with this stuff (`getDynamicPlacholders`, `setDynamicPlacholders`), a type hinting in appropriate `View` methods and a some documentation for other developers.
Of course if it's welcome. ;)

**Review by dynasource — APPROVED** [review]

**Comment by samdark** [comment]
Merged. Thank you!

**Pull request merged**